command/agent/cache/lease_cache.go

```diff
@@ -0,0 +1,813 @@
+		return
+	}
+
+	c.logger.Debug("initiating renewal", "path", req.Request.URL.Path, "method", req.Request.Method)
+	go renewer.Renew()
+	defer renewer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			// This is the case which captures context cancellations from token
+			// and leases. Since all the contexts are derived from the agent's
+			// context, this will also cover the shutdown scenario.
+			c.logger.Debug("context cancelled; stopping renewer", "path", req.Request.URL.Path)
+			return
+		case err := <-renewer.DoneCh():
+			// This case covers renewal completion and renewal errors
+			if err != nil {
+				c.logger.Error("failed to renew secret", "error", err)
+				return
+			}
+			c.logger.Debug("renewal halted; evicting from cache", "path", req.Request.URL.Path)
+			return
+		case renewal := <-renewer.RenewCh():
+			// This case captures secret renewals. Renewed secret is updated in
+			// the cached index.
+			c.logger.Debug("renewal received; updating cache", "path", req.Request.URL.Path)
+			err = c.updateResponse(ctx, renewal)
+			if err != nil {
+				c.logger.Error("failed to handle renewal", "error", err)
+				return
+			}
+		case <-index.RenewCtxInfo.DoneCh:
+			// This case indicates the renewal process to shutdown and evict
+			// the cache entry. This is triggered when a specific secret
+			// renewal needs to be killed without affecting any of the derived
+			// context renewals.
+			c.logger.Debug("done channel closed")
+			return
+		}
+	}
+}
+
+func (c *LeaseCache) updateResponse(ctx context.Context, renewal *api.RenewOutput) error {
+	id := ctx.Value(contextIndexID).(string)
+
+	// Get the cached index using the id in the context
+	index, err := c.db.Get(cachememdb.IndexNameID, id)
+	if err != nil {
+		return err
+	}
+	if index == nil {
+		return fmt.Errorf("missing cache entry for id: %q", id)
+	}
+
+	// Read the response from the index
+	resp, err := http.ReadResponse(bufio.NewReader(bytes.NewReader(index.Response)), nil)
+	if err != nil {
+		c.logger.Error("failed to deserialize response", "error", err)
+		return err
+	}
+
+	// Update the body in the reponse by the renewed secret
+	bodyBytes, err := json.Marshal(renewal.Secret)
+	if err != nil {
+		return err
+	}
+	if resp.Body != nil {
+		resp.Body.Close()
+	}
+	resp.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
+	resp.ContentLength = int64(len(bodyBytes))
+
+	// Serialize the response
+	var respBytes bytes.Buffer
+	err = resp.Write(&respBytes)
+	if err != nil {
+		c.logger.Error("failed to serialize updated response", "error", err)
+		return err
+	}
+
+	// Update the response in the index and set it in the cache
+	index.Response = respBytes.Bytes()
+	err = c.db.Set(index)
+	if err != nil {
+		c.logger.Error("failed to cache the proxied response", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// computeIndexID results in a value that uniquely identifies a request
+// received by the agent. It does so by SHA256 hashing the serialized request
```

Comment:
I recommend using Blake2 instead of SHA256. You don't need the security and Blake is way faster.

Reply:
Agreed. Fixing in https://github.com/hashicorp/vault/pull/6258/commits/f31ebb37c5d6aea26d0d9ad563d98d64aa5c65bb.